src/main/java/seedu/address/history/CommandHistory.java

```diff
@@ -0,0 +1,159 @@
+package seedu.address.history;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+
+import seedu.address.history.exception.HistoryException;
+
+public class CommandHistory implements History {
+    private static final String MESSAGE_LENGTH_LIMIT_AT_LEAST_ONE = "Length of Command History must at least be 1";
+
+    private List<String> commandHistory;
+    private int currentCommandIndex;
+    private final int lengthLimit;
+    private boolean hasReturnedCurrentCommandBefore;
+
+    /**
+     * Constructor for CommandHistory
+     * @param lengthLimit length limit of command history. Must be strictly more than 0.
+     * @throws HistoryException if lengthLimit is less than or equals to 0.
+     */
+    public CommandHistory(int lengthLimit) throws HistoryException {
+        if (lengthLimit > 0) {
+            this.commandHistory = new ArrayList<>(lengthLimit);
+
+            // length limit is one-indexed similar to commandHistory.size()
+            this.lengthLimit = lengthLimit;
+
+            // since commandHistory is zero-indexed, initializing currentCommandIndex to -1 solves the zero-index issue
+            this.currentCommandIndex = -1;
+
+            this.hasReturnedCurrentCommandBefore = false;
+        } else {
+            throw new HistoryException(MESSAGE_LENGTH_LIMIT_AT_LEAST_ONE);
+        }
+    }
+
+    public int getCurrentCommandIndex() {
+        return currentCommandIndex;
+    }
+
+    public int getLengthLimit() {
+        return lengthLimit;
+    }
+
+    public List<String> getCommandHistory() {
+        return commandHistory;
+    }
+
+    private boolean isLimitReached() {
+        return commandHistory.size() == lengthLimit;
+    }
+
+    private boolean isWithinLengthLimit() {
+        return currentCommandIndex >= 0 && currentCommandIndex < lengthLimit;
+    }
+
+    private boolean isAbleToReturnNextCommand() {
+        return currentCommandIndex < commandHistory.size() - 1 && isWithinLengthLimit();
+    }
+
+    private boolean isAbleToReturnPreviousCommand() {
+        return currentCommandIndex > 0 && isWithinLengthLimit();
+    }
+
+    private boolean isCurrentCommandIndexZero() {
```

Comment:
Is this abstraction really necessary?

Reply:
to adhere to similar level of abstraction in the code

Reply:
Fair enough